### ISSUE TITLE
`Tera::render_str(...)` requiring `&mut self` is unintuitive, ought to be fine with `&self`

### ISSUE THREAD

**Issue by jokeyrhyme** [Keats/tera#750]
Howdie, thanks so much for sharing this project <3

I noticed today that we have a very useful (to me) function here: [`Tera::render_str(&mut self, ...) -> ...`](https://github.com/Keats/tera/blob/cbd790e218a71067aed6a1a59ace97e78ca35c38/src/tera.rs#L355-L360)

However, it requires `&mut self` which seems a little odd at first because (at least for me) all of the other render methods are happy with `&self`

I can see why this is necessary from the way the library works internally, so it does make sense after reading the code

I wonder if there's any desire to allow one-off renders to work without needing them to be stored within the `Tera` struct first?

Cheers!

**Comment by Keats** [comment]
You can use Tera::one_off for that: https://github.com/Keats/tera/blob/cbd790e218a71067aed6a1a59ace97e78ca35c38/src/tera.rs#L374 but you don't get custom filters/fn etc

**Comment by jokeyrhyme** [comment]
Yeah, I did consider that, but for my use case I do still need custom filters to work

My current work around is to use a `std::sync::Mutex` to ensure exclusivity, and performance isn't a killer for this use case, so I wouldn't say this is urgent for me

**Comment by chriskrycho** [comment]
I was poking at this same thing, and I was wondering if you'd be open to adding something like a `render_one_off()` (name to be improved!) which takes `&self` instead of `&mut self` to allow thread safe sharing, and trades that off for not being able to participate in things like the inheritance chain. My broad strokes use case is being able to basically "pre-load" all of the foundational templates for a given page, then pass a single Tera instance by reference to a bunch of different threads so that they can *use* that pre-loaded template information, but they do not need to *change it* in any way, and thereby also avoid instantiating and thus allocating for every consumer.

In my specific use case, I want to pre-process a bunch of Markdown with embedded Tera invocations, but you can imagine the same thing being useful in many other similar cases, where the content to be rendered wants access to the set of underlying templates but does not need to be able to (indeed, in my case, *should not*) change anything about the baseline once loaded.

This comes at the cost of not supporting adding to the inheritance chain, but also means that all existing filters etc. should "just work", and dodges a *lot* of allocations!

Happy to open a PR if you are interested; I already implemented it and am using it, but I also understand if you are *not* interested and am happy to just keep it in my own fork.

**Comment by Keats** [comment]
I'm interested in seeing the diff but I'm not planning to merge any of that in v1. [v2](https://github.com/Keats/tera2) uses bytecode and is optimizing some stuff like macro lookup in a way where you still do have to inheritance.
The functions/filters/tests are not implemented yet (looking for more feedback on how to improve them!) but should happen soonish, at least in a first version.

**Comment by Keats** [comment]
Closing as part of the Tera v2 release. If this is still relevant in v2, please reopen — see the 
  migration guide: https://github.com/Keats/tera/blob/master/MIGRATION.md
